daemon/protocol/client_lib.py

```diff
@@ -0,0 +1,228 @@
+    """
+    """
+    global _client
+    if not is_started():
+        raise ScootExceptoin("Not started.")
+    req = daemon_pb2.CheckoutSnapshotRequest(snapshot_id=snapshot_id, dir=dirpath)
+    try: resp = _client.CheckoutSnapshot(req)
+    except Exception as e: raise ScootException(e)
+    if resp.error != "":
+        raise ScootException(resp.error)
+    return
+
+
+def run(snapshot_id, argv, env=None, timeout_ns=0, src_paths_to_dest_dirs=None, ):
+    """
+    """
+    global _client
+    if not is_started():
+        raise ScootExceptoin("Not started.")
+    cmd = daemon_pb2.RunRequest.Command(timeout_ns=timeout_ns)
+    cmd.argv.extend(argv)
+    for key, val in env.iteritems():
+        cmd.env[key] = val
+
+    plan = daemon_pb2.RunRequest.OutputPlan()
+    for key, val in src_paths_to_dest_dirs.iteritems():
+        plan.src_paths_to_dest_dirs[key] = val
+
+    req = daemon_pb2.RunRequest(snapshot_id=snapshot_id, cmd=cmd, plan=plan)
+    try: resp = _client.Run(req)
+    except Exception as e: raise ScootException(e)
+    if resp.error != "":
+        raise ScootException(resp.error)
+    return resp.run_id
+
+
+def poll(run_ids, timeout_ns=0, return_all=False):
+    """
+    """
+    global _client
+    if not is_started():
+        raise ScootExceptoin("Not started.")
+    req = daemon_pb2.PollRequest(timeout_ns=timeout_ns, all=return_all)
+    req.run_ids.extend(run_ids)
+    try: resp = _client.Poll(req)
+    except Exception as e: raise ScootException(e)
+
+    def domain(status):
+        return ScootStatus(run_id=status.run_id,
+                           state=status.state,
+                           snapshot_id=status.snapshot_id,
+                           exit_code=status.exit_code,
+                           error=status.error)
+    return [domain(x) for x in resp.status]
+
+
+if __name__ == '__main__':
+    # Adapted from daemon/server/handler_test.go
+    # TODO: make this a proper unit test.
+    # TODO: use a tmp domain socket instead of the default.
+    start()
+    tmpdir = tempfile.mkdtemp()
+    try:
+        # Populate the paths we want to ingest.
+	script = os.path.join(tmpdir, "script.sh")
+	resource = os.path.join(tmpdir, "resource.txt")
+        open(script, 'w').write("#!/bin/sh\nls resource.txt")
+        open(resource, 'w').write("content")
+
+	# Ingest scripts into their own snapshots. The 'fail' snapshot will be missing resource.txt.
+        ok_id = create_snapshot(os.path.join(tmpdir, "*"))
+        fail_id = create_snapshot(script)
+
+	# Run scripts serially in their respective snapshots. Block until each run finishes.
+	ok_run_id = run(argv=["./script.sh"], timeout_ns=500*1e6, snapshotId=ok_id)
+        ok_statuses = poll(run_ids=[ok_run_id], timeout_ns=500*1e6)
+        if len(ok_statuses) != 1:
+            raise ScootException("expected one poll result for ok_run_id.")
+
+	fail_run_id = run(argv=["./script.sh"], timeout_ns=500*1e6, snapshotId=fail_id)
```

Comment:
+1 - it'd be great to use 2 space indent too (vs 4, currently) to match Twitter/Pants/Google/etc style.